### PR TITLE
[Bounty][Feature]Department Heads can Add Access to their own departments to IDs

### PR DESCRIPTION
**Pull request by KirkwallDay**
## About The Pull Request
This pull reworks the ID computer program so department heads (Or anyone with the sub-access ID change) can add access only to specific departments. This was requested on incentivized orders. Here's a breakdown of the changes:

- Added a sub-access for each department to change ID's. Heads start out with these sub-access' and they can add them to anyone else. These new ID computer accesses only work within the individual department they're for, AND General area. The Club Manager can only edit general access.
- The original ID Computer access has been changed to Global ID Access, and is still under command.
- Changed maintenance tunnels from Engineering area to General Area, as I think anyone should be able to give out maint access.
- Redid some of the ID computer functions, any sub-access can also edit card names, account numbers, etc...
- Only the Global ID Access can terminate an employee.
- The job buttons mostly don't work for sub-access, as most of the jobs have access to places outside of their department.

## Why It's Good For The Game
If a department head or someone they delegate wants to hire a vagabond or allow someone else into their department. They can just do it, instead of pestering the Captain or FO.

## Changelog
:cl: yorpyorpyorp
add: Sub-access types to ID modification. Department heads can now add or remove access to their departments and general ship areas.
tweak: re-did some of the ID computer code, and moved a few access types around. Maint is now a general access type, instead of an engineering one.
refactor: Refactored the authorization function, to allow it to check for all the ID mod subtypes.
/:cl:

### CODE REVIEW

**Review by genessee596 — APPROVED** [review]
Delegate stamp of approval.

Not sure why this isn't a feature already - should take a lot of confusion out of job changes when the head of the department can hire new subordinates instead of sending them to the FO.

File: code/modules/modular_computers/file_system/programs/command/card.dm [hunk 1]
```diff
@@ -118,7 +127,8 @@
 	var/obj/item/weapon/card/id/id_card
 	if (computer.card_slot)
 		id_card = computer.card_slot.stored_card
-	if (!user_id_card)
+	if (!user_id_card || !authorized(user_id_card))
+		to_chat(user, "<span_class='warning'>Access denied</span>")
```
Comment (mattroks101) [inline code/modules/modular_computers/file_system/programs/command/card.dm:131]: Use the span macros please.

File: code/modules/modular_computers/file_system/programs/command/card.dm [hunk 2]
```diff
@@ -233,8 +243,11 @@
 						if(!jobdatum)
 							to_chat(usr, "<span class='warning'>No log exists for this job: [t1]</span>")
 							return
-
 						access = jobdatum.get_access()
+						for(var/A in access)
+							if(!check_modify(user_id_card, A))
+								to_chat(user, "<span class='warning'>Access denied</span>")
```
Comment (mattroks101) [inline code/modules/modular_computers/file_system/programs/command/card.dm:249]: Same here.

File: code/modules/modular_computers/file_system/programs/command/card.dm [hunk 3]
```diff
@@ -247,13 +260,13 @@
...
-							id_card.access -= access_type
-							if(!access_allowed)
-								id_card.access += access_type
-							break
+					if (check_modify(user_id_card, access_type))
+						id_card.access -= access_type
+						if(!access_allowed)
+							id_card.access += access_type
+					else
+						to_chat(user, "<span_class='warning'>Access denied</span>")
```
Comment (mattroks101) [inline code/modules/modular_computers/file_system/programs/command/card.dm:268]: And here.

File: code/modules/modular_computers/file_system/programs/command/card.dm [hunk 4]
```diff
@@ -233,8 +243,11 @@
 						if(!jobdatum)
 							to_chat(usr, "<span class='warning'>No log exists for this job: [t1]</span>")
```
Comment (mattroks101) [inline code/modules/modular_computers/file_system/programs/command/card.dm:244]: I know that the other spans in here here don't use the span macros, but if you're adding new ones they should use the span macros. I'm not going to ask you to clean up the whole file and add span macros for everything unless you really want to, just that new things being added to the codebase use the correct format.

**Review by mattroks101 — DISMISSED** [review]

File: code/modules/modular_computers/file_system/programs/command/card.dm (hunk 4 above)
Comment (KirkwallDay) [reply code/modules/modular_computers/file_system/programs/command/card.dm:244]: No problem, I'll change my code to use the macros. Didn't actually know about them until you mentioned it.

Comment (KirkwallDay) [reply code/modules/modular_computers/file_system/programs/command/card.dm]: Okay, I went ahead and changed all the span code in the file to use macros instead. It was a simple find/replace.

**Review by TorinoFermic — APPROVED** [review]

**Pull request merged**